assistant/assistant.go

```diff
@@ -0,0 +1,105 @@
+package assistant
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/influxdata/telegraf/config"
+	"golang.org/x/net/websocket"
+)
+
+// Assistant represents a client that will connect to the cloud server to allow
+// for plugin management remotely.
+type Assistant struct {
+	Config *config.Config // stores plugins and agent config
+}
+
+// NewAssistant returns an Assistant for the given Config.
+func NewAssistant(config *config.Config) (*Assistant, error) {
+	a := &Assistant{
+		Config: config,
+	}
+	return a, nil
+}
+
+// client stores the channels to communicate with the agent and the websocket
+// connection to the cloud server.
+// TODO determine if Ctx is necessary, and if so, what for
+type client struct {
```

Comment:
Not too convinced that this intermediary client is really needed. What benefit is it providing over just writing everything as part of the Assistant directly? Seems like it's making it somewhat harder to deal with things as all you can do from the Agent's point of view is send and receive bytes over two channels, and nothing else. 

What are your thoughts on a design more like:
The Assistant takes care of all the connections and exposes two things:
1. A function called Write which is responsible for sending messages to the cloud
- Takes in like a status (success, failure) etc.. as well as a UUID (what it's responding to), and any payload data
2. Exposes a channel of maybe `AssistantCommand` objects or something, which the Agent listens on, takes action when something is passed in through the channel, then writes the response to the Assistant

An `AssistantCommand` can have fields like `CommandType` which could be like `START`, `STOP`, `UPDATE` as well as other things like what plugins is it talking about, configs etc..

Reply:
Basically my idea is the Assistant is just the intermediary that handles the connection between the Agent and the Cloud, making it easier for the Agent by doing all the marshalling/unmarshalling of objects into easy to use types.

Reply:
Let me know what you think

Reply:
True this client struct does seem like an extra struct. Can just put all the functionality into assistant and be ok.

Reply:
I liked Dave's idea about having the two responsibilities separated.

The Agent can expose functions such as `StartPlugin()`, `StopPlugin()`, `UpdatePlugin` etc... and then the assistant does all the communication internally, calling those Agent plugins and returning to the cloud server the responses and whether it was a success or failure.